File: Week-01/Day-05/main.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{38, 27, 43, 10}
	fmt.Println("Given Array: ", arr)

	// Merge Sort
	mSort := mergeSort(arr)
	fmt.Println("Merge Sort: ", mSort)

	// Quick Sort
	fmt.Println("Quick Sort: ", quickSort(arr))

	// heap sort
	fmt.Println("Heap Sort: ", heapSort(arr))
}

// Merge Sort
func mergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := mergeSort(arr[:mid])
	right := mergeSort(arr[mid:])
	return merge(left, right)
}

func merge(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))
	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

// Quick Sort
func quickSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}
	left, right := 0, len(arr)-1
	pivotIndex := len(arr) / 2
	arr[pivotIndex], arr[right] = arr[right], arr[pivotIndex]

	for i := range arr {
		if arr[i] < arr[right] {
			arr[i], arr[left] = arr[left], arr[i]
			left++
		}
	}
	arr[left], arr[right] = arr[right], arr[left]

	quickSort(arr[:left])
	quickSort(arr[left+1:])
	return arr
}

// Heap Sort
func heapSort(arr []int) []int {
	n := len(arr)

	for i := n/2 - 1; i >= 0; i-- {
		heapify(arr, n, i)
	}

	for i := n - 1; i >= 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapify(arr, i, 0)
	}

	return arr
}

func heapify(arr []int, n, i int) {
	largest := i
	left := 2*i + 1
	right := 2*i + 2

	if left < n && arr[left] > arr[largest] {
		largest = left
	}

	if right < n && arr[right] > arr[largest] {
		largest = right
	}

	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)
	}
}

File: Week-01/Day-05/main.cpp
#include <iostream>
using namespace std;

void heapify(int arr[], int n, int i){
  int largest = i; // Initialize largest as root 
  int left = 2 * i + 1; // left child
  int right = 2 * i + 2; // right child 
  
  // If left child is larger than root 
  if(left < n && arr[left] > arr[largest])
    largest = left; 

  // If right child is larger than largest so far
  if (right < n && arr[right] > arr[largest])
    largest = right;

  // If largest is not root 
  if (largest != i) {
    swap(arr[i], arr[largest]);

    // Recursively heapify the affected sub-tree
    heapify(arr, n, largest);
  }
}

void heapSort(int arr[], int n){
  // Build heap(rearrange array)
  for(int i = n/2-1; i >=0; i--){
    heapify(arr, n, i); 

  // One by one extract elements from heap 
  for(int i = n-1; i >= 0; i--){
      // Move current root to end 
      swap(arr[0], arr[i]);

      //call max heapify on the reduced heap 
      heapify(arr, i, 0 );
    }
  }
}
